File: WinPaletter/GUI/Elements/Separators/SeparatorH.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D

Namespace UI.WP

    <Description("Horizontal separator for WinPaletter UI")>
    Public Class SeparatorH : Inherits Control

        Sub New()
            TabStop = False
            DoubleBuffered = True
            Text = ""
        End Sub

        <Browsable(True)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)>
        <EditorBrowsable(EditorBrowsableState.Always)>
        <Editor(GetType(System.ComponentModel.Design.MultilineStringEditor), GetType(System.Drawing.Design.UITypeEditor))>
        <Bindable(True)>
        Public Overrides Property Text As String = ""

        Public Property AlternativeLook As Boolean = False

        Protected Overrides Sub OnResize(e As EventArgs)
            MyBase.OnResize(e)
            Size = New Size(Width, If(Not AlternativeLook, 1, 2))
        End Sub

        Protected Overrides Sub OnPaint(e As PaintEventArgs)
            Dim G As Graphics = e.Graphics
            G.SmoothingMode = SmoothingMode.AntiAlias
            DoubleBuffered = True
            MyBase.OnPaint(e)

            Dim clr As Color
            clr = Color.FromArgb(75, 75, 75)

            '################################################################################# Customizer
            Dim IdleLine As Color

            If Parent IsNot Nothing Then

                If My.Style.DarkMode Then
                    If Not AlternativeLook Then
                        IdleLine = Parent.BackColor.CB(0.1)
                    Else
                        IdleLine = Color.DarkRed
                    End If
                Else
                    If Not AlternativeLook Then
                        IdleLine = Parent.BackColor.CB(-0.1)
                    Else
                        IdleLine = Color.DarkRed
                    End If
                End If

            Else
                If My.Style.DarkMode Then IdleLine = Color.FromArgb(76, 76, 76) Else IdleLine = Color.FromArgb(210, 210, 210)
            End If

            '################################################################################# Customizer

            Using C As New Pen(IdleLine, If(Not AlternativeLook, 1, 2))
                G.DrawLine(C, New Point(0, 0), New Point(Width, 0))
                G.DrawLine(C, New Point(0, 1), New Point(Width, 1))
            End Using
        End Sub

    End Class

End Namespace

File: WinPaletter/GUI/Elements/Separators/SeparatorH.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Drawing.Design;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

namespace WinPaletter.UI.WP
{
    [Description("Horizontal separator for WinPaletter UI")]
    public class SeparatorH : Control
    {
        public SeparatorH()
        {
            SetStyle(ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor, true);
            DoubleBuffered = true;
            BackColor = Color.Transparent;
            TabStop = false;
            Text = string.Empty;
        }

        #region Properties

        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        [Editor(typeof(MultilineStringEditor), typeof(UITypeEditor))]
        [Bindable(true)]
        public override string Text { get; set; } = string.Empty;

        public bool AlternativeLook { get; set; } = false;

        #endregion

        #region Events/Overrides

        protected override void OnResize(EventArgs e)
        {
            Size = new(Width, !AlternativeLook ? 1 : 2);

            base.OnResize(e);
        }

        int parentLevel = 0;
        protected override void OnParentChanged(EventArgs e)
        {
            base.OnParentChanged(e);

            parentLevel = this.Level();
        }


        #endregion

        protected override void OnPaintBackground(PaintEventArgs pevent)
        {
            //Leave it empty to make control background transparent
            base.OnPaintBackground(pevent);
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics G = e.Graphics;
            G.SmoothingMode = SmoothingMode.AntiAlias;

            //Makes background drawn properly, and transparent
            InvokePaintBackground(this, e);

            Config.Scheme scheme = Enabled ? Program.Style.Schemes.Main : Program.Style.Schemes.Disabled;

            Color Line = AlternativeLook ? Color.DarkRed : Color.FromArgb(50, scheme.Colors.ForeColor);

            using (Pen P = new(Line, !AlternativeLook ? 1 : 2))
            {
                G.DrawLine(P, new Point(0, 0), new Point(Width, 0));
                G.DrawLine(P, new Point(0, 1), new Point(Width, 1));
            }

            base.OnPaint(e);


        }
    }
}